src/Documents/Conventions/DocumentConventions.ts

```diff
@@ -179,20 +185,23 @@ export class DocumentConventions {
     return Serializer.toJSON<T>(document, document['@metadata'] || {});
   }
 
-  public setIdOnDocument<T extends Object = IRavenObject>(document: T, id: string, documentType?: DocumentType<T>): T {
-    const docType: DocumentType<T> = documentType || this.getTypeFromDocument(document);
+  public setIdOnDocument<T extends Object = IRavenObject>(document: T, id: string, options: IGetSetIDOnDocumentOptions<T> = {}): T {
+    let docType: DocumentType<T> = options.documentType || this.getTypeFromDocument(document);
     const idProperty = this.getIdPropertyName(docType, document);
 
     if ('object' !== (typeof document)) {
       throw new InvalidOperationException("Invalid entity provided. It should implement object interface");
     }
 
-    document[idProperty] = id;
+    if ((false !== options.setIdIfPropertyIsNotDefined) || (idProperty in document)) {
```

Comment:
if javascript object literal  does not have an id property defined we don't have anything to populate.